chunked_stream/lib/src/chunked_stream_iterator.dart

```diff
@@ -23,197 +23,179 @@ import 'dart:math';
+  /// automatically drained.
+  @override
   Stream<List<T>> substream(int size) {
-    ArgumentError.checkNotNull(size, 'size');
-    if (size <= 0) {
-      throw ArgumentError.value(size, 'size', 'size must be greater than zero');
-    }
-    if (_wakeup != null) {
-      throw StateError('Concurrent invocations not supported');
-    }
+    _toRead = size;
+    _substreamId++;
 
-    final c = StreamController<List<T>>(
-      onListen: _subscription.resume,
-      onPause: _subscription.pause,
-      onResume: _subscription.resume,
-      onCancel: () async {
-        // Read the rest of size bytes
-        final c = Completer<void>();
-        _wakeup = () {
-          if (_lastChunk != null) {
-            if (_lastChunk.length < size) {
-              size -= _lastChunk.length;
-            } else {
-              _buffered.addAll(_lastChunk.sublist(size));
-              size = 0;
-            }
-            _lastChunk = null;
-          }
-          if (size <= 0) {
-            _reset();
-            c.complete();
-            return;
-          }
-          if (_error != null) {
-            // Ignore errors, as we're skipping in the stream
-            _error = null;
-            _stackTrace = null;
-          }
-          if (_isDone) {
-            _reset();
-            c.complete();
-            return;
-          }
-        };
-        _wakeup();
-        await c.future;
-      },
-    );
-    if (_buffered.isNotEmpty) {
-      final n = min(size, _buffered.length);
-      c.add(_buffered.sublist(0, n));
-      _buffered = _buffered.sublist(n);
-      size -= n;
-    }
-    _wakeup = () {
-      if (_lastChunk != null) {
-        if (_lastChunk.length < size) {
-          c.add(_lastChunk);
-          size -= _lastChunk.length;
-        } else {
-          c.add(_lastChunk.sublist(0, size));
-          _buffered.addAll(_lastChunk.sublist(size));
-          size = 0;
+    return _substream(_substreamId);
+  }
+
+  Stream<List<T>> _substream(int substreamId) async* {
+    /// Only yield when we are dealing with the same substream and there are
+    /// elements to be read.
+    while (_substreamId == substreamId && _toRead > 0) {
+      if (_buffered.isEmpty) {
+        if (!(await _iterator.moveNext())) {
+          break;
         }
-        _lastChunk = null;
-      }
-      assert(size >= 0, 'size should always be positive');
-      if (size == 0) {
-        _reset();
-        c.close();
-        return;
-      }
-      if (_error != null) {
-        c.addError(_error, _stackTrace);
-        _error = null;
-        _stackTrace = null;
-        return;
-      }
-      if (_isDone) {
-        _reset();
-        c.close();
-        return;
+
+        _buffered = _iterator.current;
       }
-    };
-    _wakeup();
 
-    return c.stream;
-  }
+      final yieldBlockEnd = min(_toRead, _buffered.length);
+      final toYield = _buffered.sublist(0, yieldBlockEnd);
+      _buffered = _buffered.sublist(yieldBlockEnd);
```

Comment:
We could special case `_toRead < _buffered.length` here too..
```dart
_buffered = _toRead < _buffered.length ? buffered.sublist(_toRead) : _emptyList;
```

Then have an instance field:
```dart
final _emptyList = <T>[];
```
That way we don't  create a new empty list object whenever we need one... (we never mutate the `_buffered` list, so we can just use a special object to represent the empty list). This might even be better than using `null`, because it allows the compiler to skip `null` checks, if we initialize `_buffered = _emptyList`.